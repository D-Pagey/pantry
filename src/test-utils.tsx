import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';
import React, { FC, ReactNode } from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { FirebaseContext } from './components/ProviderFirebase';

type AllTheProvidersProps = {
    children: ReactNode;
    customContext?: any;
};

const AllTheProviders: FC<AllTheProvidersProps> = ({ children, customContext }) => (
    <FirebaseContext.Provider
        value={{
            categories: [],
            deleteFoodItem: () => () => null,
            expiringCount: 0,
            isAuthed: false,
            isCheckingAuth: false,
            fridge: [],
            signOut: () => null,
            updateFridge: () => null,
            user: {
                email: null,
                name: null
            },
            ...customContext
        }}
    >
        <MemoryRouter>{children}</MemoryRouter>
    </FirebaseContext.Provider>
);

const customRender = (ui: any, customContext?: any, options?: any) =>
    render(ui, { wrapper: () => <AllTheProviders customContext={customContext}>{ui}</AllTheProviders>, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
