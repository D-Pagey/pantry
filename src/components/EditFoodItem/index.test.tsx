import React from 'react';
import userEvent from '@testing-library/user-event';
import { render } from '../../test-utils';
import { Fridge } from '../../fixtures';
import { EditFoodItem } from '.';

const props = {
    item: Fridge[0]
};

describe('EditFoodItem component', () => {
    it('should render', () => {
        const { container } = render(<EditFoodItem {...props} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should update remove correct servings amount', () => {
        const updateFridge = jest.fn();
        const { getByTestId } = render(<EditFoodItem {...props} />, { updateFridge });

        userEvent.click(getByTestId('2-0'));
        userEvent.click(getByTestId('3-0'));
        userEvent.click(getByTestId('editFoodItemSubmit'));

        expect(updateFridge).toHaveBeenCalledWith({
            batches: [
                { expires: expect.any(Date), owner: '1', servings: 1 },
                { expires: expect.any(Date), owner: '2', servings: 1 },
                { expires: expect.any(Date), owner: '3', servings: 3 }
            ],
            category: 'vegetables',
            name: 'carrots'
        });
    });

    it('should remove a batch if checked last serving', () => {
        const updateFridge = jest.fn();
        const { getByTestId } = render(<EditFoodItem {...props} />, { updateFridge });

        userEvent.click(getByTestId('1-0'));
        userEvent.click(getByTestId('editFoodItemSubmit'));

        expect(updateFridge).toHaveBeenCalledWith({
            batches: [
                { expires: expect.any(Date), owner: '2', servings: 2 },
                { expires: expect.any(Date), owner: '3', servings: 4 }
            ],
            category: 'vegetables',
            name: 'carrots'
        });
    });

    it('should update remove correct servings amount when checked and unchecked', () => {
        const updateFridge = jest.fn();
        const { getByTestId } = render(<EditFoodItem {...props} />, { updateFridge });

        userEvent.click(getByTestId('1-0'));
        userEvent.click(getByTestId('1-0'));

        userEvent.click(getByTestId('2-0'));
        userEvent.click(getByTestId('2-0'));

        userEvent.click(getByTestId('3-0'));
        userEvent.click(getByTestId('editFoodItemSubmit'));

        expect(updateFridge).toHaveBeenCalledWith({
            batches: [
                { expires: expect.any(Date), owner: '1', servings: 1 },
                { expires: expect.any(Date), owner: '2', servings: 2 },
                { expires: expect.any(Date), owner: '3', servings: 3 }
            ],
            category: 'vegetables',
            name: 'carrots'
        });
    });
});
