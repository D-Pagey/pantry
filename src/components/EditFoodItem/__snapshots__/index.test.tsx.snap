// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditFoodItem component should render 1`] = `
.c6 {
  background-color: #348841;
  border: 0;
  border-radius: 5px;
  color: white;
  cursor: pointer;
  padding: 1rem;
  min-width: 68px;
  background-color: white;
  border: 1px solid #348841;
  color: #348841;
}

.c5 {
  background-color: #348841;
  border: 0;
  border-radius: 5px;
  color: white;
  cursor: pointer;
  padding: 1rem;
  margin: 0 0 1rem;
  min-width: 68px;
  background-color: #B2B2B2;
  color: white;
  cursor: not-allowed;
  margin: 0 0 1rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  font-weight: normal;
  margin: 0 0 1.5rem;
}

.c2 {
  margin: 0 0 2rem;
  padding: 0;
  list-style: none;
}

.c3 {
  border-bottom: 1px solid #E5E5E5;
  padding: 1rem 0 1rem 3rem;
}

.c3:first-child {
  border-top: 1px solid #E5E5E5;
}

.c4 {
  margin: 0 1rem 0 0;
}

.c4 .rc-checkbox.inner {
  border-color: red;
}

<div
  class="c0"
>
  <h4
    class="c1"
  >
    How many 
    carrots
     servings are you eating?
  </h4>
  <ul
    class="c2"
  >
    <li
      class="c3"
    >
      <span
        class="rc-checkbox c4"
      >
        <input
          class="rc-checkbox-input"
          data-testid="1-0"
          type="checkbox"
          value=""
        />
        <span
          class="rc-checkbox-inner"
        />
      </span>
      Expired 
      24th Jun
    </li>
    <li
      class="c3"
    >
      <span
        class="rc-checkbox c4"
      >
        <input
          class="rc-checkbox-input"
          data-testid="2-0"
          type="checkbox"
          value=""
        />
        <span
          class="rc-checkbox-inner"
        />
      </span>
      Expired 
      26th Jun
    </li>
    <li
      class="c3"
    >
      <span
        class="rc-checkbox c4"
      >
        <input
          class="rc-checkbox-input"
          data-testid="2-1"
          type="checkbox"
          value=""
        />
        <span
          class="rc-checkbox-inner"
        />
      </span>
      Expired 
      26th Jun
    </li>
    <li
      class="c3"
    >
      <span
        class="rc-checkbox c4"
      >
        <input
          class="rc-checkbox-input"
          data-testid="3-0"
          type="checkbox"
          value=""
        />
        <span
          class="rc-checkbox-inner"
        />
      </span>
      Expired 
      29th Jun
    </li>
    <li
      class="c3"
    >
      <span
        class="rc-checkbox c4"
      >
        <input
          class="rc-checkbox-input"
          data-testid="3-1"
          type="checkbox"
          value=""
        />
        <span
          class="rc-checkbox-inner"
        />
      </span>
      Expired 
      29th Jun
    </li>
    <li
      class="c3"
    >
      <span
        class="rc-checkbox c4"
      >
        <input
          class="rc-checkbox-input"
          data-testid="3-2"
          type="checkbox"
          value=""
        />
        <span
          class="rc-checkbox-inner"
        />
      </span>
      Expired 
      29th Jun
    </li>
    <li
      class="c3"
    >
      <span
        class="rc-checkbox c4"
      >
        <input
          class="rc-checkbox-input"
          data-testid="3-3"
          type="checkbox"
          value=""
        />
        <span
          class="rc-checkbox-inner"
        />
      </span>
      Expired 
      29th Jun
    </li>
  </ul>
  <button
    class="sc-AxjAm c5"
    data-testid="editFoodItemSubmit"
    disabled=""
    type="button"
  >
    Eat 
    0
     Carrots
  </button>
  <button
    class="sc-AxjAm c6"
    type="button"
  >
    Cancel
  </button>
</div>
`;
